File: elixir/rover/test/rover_test.exs
defmodule RoverTest do
  use ExUnit.Case, async: true
  doctest Rover

  describe "at" do
    test "makes a rover at the position and direction" do
      rover = Rover.at({1, 2}, :N)
      assert rover.position == {1, 2}
      assert rover.direction == :N
    end
  end

  describe "status" do
    test "returns a string with the current position and direction" do
      assert Rover.status(Rover.at({1, 2}, :N)) == "1 2 N"
    end
  end

  test "turning left" do
    assert Rover.turn_left(Rover.at({1, 2}, :N)) == Rover.at({1, 2}, :W)
    assert Rover.turn_left(Rover.at({1, 2}, :W)) == Rover.at({1, 2}, :S)
    assert Rover.turn_left(Rover.at({1, 2}, :S)) == Rover.at({1, 2}, :E)
    assert Rover.turn_left(Rover.at({1, 2}, :E)) == Rover.at({1, 2}, :N)
  end

  test "turning right" do
    assert Rover.turn_right(Rover.at({1, 2}, :N)) == Rover.at({1, 2}, :E)
    assert Rover.turn_right(Rover.at({1, 2}, :W)) == Rover.at({1, 2}, :N)
    assert Rover.turn_right(Rover.at({1, 2}, :S)) == Rover.at({1, 2}, :W)
    assert Rover.turn_right(Rover.at({1, 2}, :E)) == Rover.at({1, 2}, :S)
  end

  test "moving forward" do
    assert Rover.move(Rover.at({1, 2}, :N)) == Rover.at({1, 3}, :N)
    assert Rover.move(Rover.at({1, 2}, :E)) == Rover.at({2, 2}, :E)
    assert Rover.move(Rover.at({1, 2}, :S)) == Rover.at({1, 1}, :S)
    assert Rover.move(Rover.at({1, 2}, :W)) == Rover.at({0, 2}, :W)
  end
end

File: kotlin/rovers/src/test/kotlin/RoverTest.kt
import org.amshove.kluent.shouldEqual
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on

internal class RoverTest : Spek({
    describe("moving forward") {
        listOf(
                Triple(Direction.NORTH, Coordinates2D(1, 2), Coordinates2D(1, 3)),
                Triple(Direction.EAST, Coordinates2D(1, 2), Coordinates2D(2, 2)),
                Triple(Direction.SOUTH, Coordinates2D(1, 2), Coordinates2D(1, 1)),
                Triple(Direction.WEST, Coordinates2D(2, 3), Coordinates2D(1, 3))
        ).forEach { (direction, initial, expected) ->
            on("a rover facing $direction") {
                val rover = Rover(initial, direction)

                it("moves from $initial to $expected") {
                    rover.moveForward()
                    rover.coordinates shouldEqual expected
                }
            }
        }
    }

    describe("turning left") {
        listOf(
                Direction.NORTH to Direction.WEST,
                Direction.WEST to Direction.SOUTH,
                Direction.SOUTH to Direction.EAST,
                Direction.EAST to Direction.NORTH
        ).forEach { (initial, expected) ->
            on("a rover facing $initial") {
                val rover = Rover(Coordinates2D(1, 2), initial)

                it("turns to $expected") {
                    rover.turnLeft()
                    rover.direction shouldEqual expected
                }
            }
        }
    }

    describe("turning right") {
        listOf(
                Direction.NORTH to Direction.EAST,
                Direction.EAST to Direction.SOUTH,
                Direction.SOUTH to Direction.WEST,
                Direction.WEST to Direction.NORTH
        ).forEach { (initial, expected) ->
            on("a rover facing $initial") {
                val rover = Rover(Coordinates2D(1, 2), initial)

                it("turns to $expected") {
                    rover.turnRight()
                    rover.direction shouldEqual expected
                }
            }
        }
    }
})
